contracts/Remittance.sol

```diff
@@ -1,16 +1,137 @@
 //SPDX-License-Identifier: MIT
 pragma solidity >= 0.6.0 < 0.7.0;
 
-import "../node_modules/@openzeppelin/contracts/access/Ownable.sol";
-import "../node_modules/@openzeppelin/contracts/utils/Pausable.sol";
+import "./access/Ownable.sol";
+import "./utils/Pausable.sol";
 
-import {SafeMath} from "../node_modules/@openzeppelin/contracts/math/SafeMath.sol";
+import {SafeMath} from "./math/SafeMath.sol";
 
 contract Remittance is Ownable, Pausable {
 
-    using SafeMath for uint;
+    using SafeMath for uint256;
 
+    mapping(bytes32 => RemittanceInstance) public remittances;
 
+    // The 'valid' element provides a flag for checking a bonafide hash key
+    struct RemittanceInstance {
+        address funder;
+        address broker;
+        uint256 fundsOwed;
+        uint256 expiryDate;
+        bool valid;
+    }
+
+    event RemittanceCreated(
+        bytes32 indexed hash,
+        address indexed funder,
+        address indexed broker,
+        uint256 amount,
+        uint256 expiryDate
+    );
+
+    event RemittanceFundsReleased(
+        bytes32 indexed hash,
+        address indexed broker,
+        uint256 amount
+    );
+
+    event RemittanceFundsReclaimed(
+        bytes32 indexed hash,
+        address indexed funder,
+        uint256 amount
+    );
+
+    // The funder can create a remittance
+    // Hash is a sha256 hash of the broker address + recipient password
+    function create(bytes32 recipientPasswordHash, address broker, uint256 expiryDate)
+    external
+    payable
+    whenNotPaused
+    returns(bool success)
+    {
+        require(broker != msg.sender, "The caller cannot be the broker");
+        require(msg.value > 0, "The amount must be greater than 0");
+
+        bytes32 rehash = keccak256ReHash(recipientPasswordHash, broker);
```

Comment:
Wait a second! You are asking the broker to pass this same hash in `release`??
How is it a difficult thing to do if that `recipientPasswordHash` is sent in the clear in `create`? 
The broker does not need to wait for the ultimate fiat recipient to even be present...
The funder might as well just pay `broker.transfer()` and be done with it.

Reply:
I think perhaps my misunderstanding.

The idea was that the funder creates a hash 'off-chain' using the password and the broker also does so once the password is provided by the recipient.

Are you saying that the broker can get the hash using a 'Man-in-the-middle attack' or is there some other record of the arguments sent to the create() function?

If that is a risk then I can change create() function to accept one single argument - a hash of the password + broker address.

A 'Man-in-the-middle attack' could use the resulting hash sent to create() to use in the reclaim() function but the call would fail because the function requires the caller to be the funder.

Reply:
> or is there some other record of the arguments sent to the create() function?

In Ethereum, all transactions are public, that includes their parameters and their execution. The broker can listen to events that mention it, then look at the transaction `.data`, and parse that to get the hash. It's not a man in the middle. It's just the rightful recipient knowing what to listen for.

The solution you describe sounds like the right one.

Reply:
Interesting. Thanks. I understand the principle now.

Reply:
I was just reviewing my question and your answer and realise that I should consider more carefully what I store in an event rather than exposing all data by default.

Reply:
Yes, an event is public too, but I want to make sure you understand that the function's parameters are public too. On a recent transaction, say here https://etherscan.io/tx/0x02e7fda12ab3d9396879ee9aef5a6c3532f13c6ebe620425421bac95202fccf0
You can see all parameters, none is secret.
So if the broker's address is not emitted in the event, a determined broker can still look at every transaction of every new block, and see if it mentions their address in the parameter. If so, they can see the hash right alongside.

Reply:
I see. Yes. Not too difficult to find and view. I can see exactly what was passed to swapTokensForExactTokens();